### ISSUE TITLE
ML

### ISSUE THREAD

**Issue by Dadabhainaoroji** [Dadabhainaoroji/Php#24]
1..Read a dataset from the user and i.) Use the Find-S algorithm to find the most specific hypothesis that is consistent with the positive examples. ii.) What is the final hypothesis after processing all the positive examples? Using the same dataset, apply the Candidate Elimination algorithm. Determine the final version space after processing all examples (both positive and negative). What are the most specific and most general hypotheses in the version space?

# FIND-S ALGORITHM

import csv

with open('enjoysport.csv', 'r') as file:

    data = [row for row in csv.reader(file)]

    print("The total number of training instances are:", len(data)-1, '\n', data[1:])

num_attribute = len(data[0]) - 1

hypothesis = ['0'] * num_attribute

for i in range(0, len(data)):

    if data[i][num_attribute] == 'yes':

        for j in range(0, num_attribute):

            if hypothesis[j] == '0' or hypothesis[j] == data[i][j]:

                hypothesis[j] = data[i][j]

            else:

                hypothesis[j] = '?'

    print("\nThe hypothesis for the training instance {} is : \n".format(i), hypothesis)

print("\nThe Maximally specific hypothesis for the training instances is ", hypothesis)

# CANDIDATE ELIMINATION ALGORITHM

import numpy as np

import pandas as pd

data = pd.read_csv('enjoysport.csv')

concepts = np.array(data.iloc[:, 0:-1])

target = np.array(data.iloc[:, -1])

def learn(concepts, target):

    specific_h = concepts[0].copy()

    general_h = [["?" for _ in range(len(specific_h))] for _ in range(len(specific_h))]

    for i, h in enumerate(concepts):

        if target[i] == "yes":

            for x in range(len(specific_h)):

                if h[x] != specific_h[x]:

                    specific_h[x] = '?'

                    general_h[x][x] = '?'

        if target[i] == "no":

            for x in range(len(specific_h)):

                if h[x] != specific_h[x]:

                    general_h[x][x] = specific_h[x]

                else:

                    general_h[x][x] = '?'

    indices = [i for i, val in enumerate(general_h) if val == ['?'] * len(specific_h)]

    for i in indices:

        general_h.remove(['?'] * len(specific_h))

    return specific_h, general_h

s_final, g_final = learn(concepts, target)

print("Final Specific_h:", s_final)

print("Final General_h:", g_final)

2..Read a dataset and use an example-based method (such as RIPPER or CN2) to generate a set of classification rules . Apply the FOIL algorithm (First-Order Inductive Learner) to learn first-order rules for predicting.

!pip install wittgenstein pandas scikit-learn

import pandas as pd

import wittgenstein as lw

from sklearn.model_selection import train_test_split

from sklearn.metrics import accuracy_score

df = pd.read_csv("playtennis.csv")

if df['PlayTennis'].dtype == object:

    df['PlayTennis'] = df['PlayTennis'].map({'Yes': True, 'No': False})

train, test = train_test_split(df, test_size=0.2, random_state=42)

ripper_clf = lw.RIPPER(random_state=42, verbosity=0)

ripper_clf.fit(train, class_feat='PlayTennis', pos_class=True)

print("Learned RIPPER Rules:")

print(ripper_clf.ruleset_)

X_test = test.drop('PlayTennis', axis=1)

y_test = test['PlayTennis']

predictions = ['Yes' if p else 'No' for p in ripper_clf.predict(X_test)]

y_test_labels = ['Yes' if y else 'No' for y in y_test]

accuracy = accuracy_score(y_test_labels, predictions)

print(f"Accuracy on test set: {accuracy:.2f}")

print("\nDetailed Rule Set Analysis:")

ripper_clf.out_model()

import pandas as pd

import wittgenstein as lw

from sklearn.model_selection import train_test_split

from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

df = pd.read_csv("mushrooms.csv")

df['class'] = df['class'].map({'e': False, 'p': True})

train, test = train_test_split(df, test_size=0.3, random_state=42)

ripper_clf = lw.RIPPER(random_state=42, verbosity=0)

ripper_clf.fit(train, class_feat='class', pos_class=True)

print("Learned RIPPER Rules:")

print(ripper_clf.ruleset_)

X_test = test.drop('class', axis=1)

y_test = test['class']

predictions = ripper_clf.predict(X_test)

accuracy = accuracy_score(y_test, predictions)

print(f"\nAccuracy on test set: {accuracy:.4f}")

print("\nDetailed Rule Set Analysis:")

ripper_clf.out_model()

print("\nConfusion Matrix:")

print(confusion_matrix(y_test, predictions))

print("\nClassification Report:")

print(classification_report(y_test, predictions, target_names=["Edible", "Poisonous"]))

3..Read a supervised dataset and use bagging and boosting technique to classify the dataset. Indicate the performance of the model.

import pandas as pd

import matplotlib.pyplot as plt

from sklearn.datasets import load_iris

from sklearn.model_selection import train_test_split

from sklearn.metrics import accuracy_score, classification_report

from sklearn.ensemble import BaggingClassifier, AdaBoostClassifier

from sklearn.tree import DecisionTreeClassifier

iris = load_iris()

X = iris.data

y = iris.target

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

bagging = BaggingClassifier(estimator=DecisionTreeClassifier(), n_estimators=50, random_state=42)

bagging.fit(X_train, y_train)

y_pred_bag = bagging.predict(X_test)

boosting = AdaBoostClassifier(estimator=DecisionTreeClassifier(max_depth=1),

                              n_estimators=50, learning_rate=1, random_state=42)

boosting.fit(X_train, y_train)

y_pred_boost = boosting.predict(X_test)

print("Bagging Performance:")

print("Accuracy:", accuracy_score(y_test, y_pred_bag))

print(classification_report(y_test, y_pred_bag, target_names=iris.target_names))

print("Boosting Performance:")

print("Accuracy:", accuracy_score(y_test, y_pred_boost))

print(classification_report(y_test, y_pred_boost, target_names=iris.target_names))

4..Read an unsupervised dataset and group the dataset based on similarity based on k-means clustering.

import pandas as pd

import matplotlib.pyplot as plt

from sklearn.cluster import KMeans

from sklearn.datasets import load_iris

iris = load_iris()

X = pd.DataFrame(iris.data, columns=iris.feature_names)

k = 3

kmeans = KMeans(n_clusters=k, random_state=42)

kmeans.fit(X)

X['cluster'] = kmeans.labels_

print("cluster centers:\n", kmeans.cluster_centers_)

plt.scatter(X.iloc[:, 0], X.iloc[:, 1], c=X['cluster'], cmap='viridis', s=40, alpha=0.6)

plt.scatter(kmeans.cluster_centers_[:, 0], kmeans.cluster_centers_[:, 1],

            c='red', marker='X', s=50, label='centroids')

plt.xlabel(iris.feature_names[0])

plt.ylabel(iris.feature_names[1])

plt.title("K-Means clustering (Iris Dataset)")

plt.legend()

plt.show()

5..Read a dataset and perform unsupervised learning using SOM algorithm.

!pip install MiniSom

import numpy as np

import matplotlib.pyplot as plt

from sklearn.datasets import load_iris

from sklearn.preprocessing import MinMaxScaler

import pandas as pd

from minisom import MiniSom

iris = load_iris()

data = iris.data

feature_names = iris.feature_names

scaler = MinMaxScaler()

scaled_data = scaler.fit_transform(data)

df = pd.DataFrame(data, columns=feature_names)

print(df.head())

som_x, som_y = 10, 10

som = MiniSom(som_x, som_y, data.shape[1], sigma=1.0, learning_rate=0.5)

som.random_weights_init(scaled_data)

som.train_random(scaled_data, 1000)

plt.figure(figsize=(10, 10))

plt.pcolor(som.distance_map().T, cmap='bone_r')

plt.colorbar()

plt.title('SOM U-Matrix Distance Map')

plt.show()

6..Write a function to generate uniform random numbers in the interval [0, 1]. Use this function to generate 10 random samples and evaluate f(x) for each sample. What are the sampled function values? Using the samples generated in the previous step, estimate the integral I using the Monte Carlo method.

import random

def uniform_random():

    return random.random()

def f(x):

    return x**2

N = 10

samples = [uniform_random() for _ in range(N)]

values = [f(x) for x in samples]

estimate = sum(values) / N

print("random sample:", samples)

print("function values:", values)

print("estimate:", estimate)

7..Read a dataset and indicate the likelihood of an event occurring using Bayesian Networks.

!pip install pgmpy

import numpy as np

import pandas as pd

from pgmpy.estimators import MaximumLikelihoodEstimator

from pgmpy.models import DiscreteBayesianNetwork

from pgmpy.inference import VariableElimination

heartDisease = pd.read_csv("heart.csv")

heartDisease = heartDisease.replace('?', np.nan)

print('Sample instances from the dataset are:')

print(heartDisease.head())

model = DiscreteBayesianNetwork([

    ('age', 'heartdisease'),

    ('sex', 'heartdisease'),

    ('exang', 'heartdisease'),

    ('cp', 'heartdisease'),

    ('heartdisease', 'restecg'),

    ('heartdisease', 'chol')

])

model.fit(heartDisease, estimator=MaximumLikelihoodEstimator)

print('Inferencing with Bayesian Network')

infer = VariableElimination(model)

print('1. Probability of HeartDisease given evidence = restecg')

q1 = infer.query(variables=['heartdisease'], evidence={'restecg': 1})

print(q1)

print('2. Probability of HeartDisease given evidence = cp')

q2 = infer.query(variables=['heartdisease'], evidence={'cp': 2})

print(q2)

Refer to the dataset in question 7 and indicate inferences based on the sequence of steps. (Same as 7)
